src/main/java/org/generationcp/middleware/service/DataImportServiceImpl.java

```diff
@@ -429,6 +430,25 @@ public void processExperimentalDesign(final Workbook workbook, final String prog
 
 	}
 
+	@Override
+	public void addEntryTypeVariableIfNotExists(final Workbook workbook,
+		final String programUUID) {
+
+		final boolean entryTypeIdExists = this.findMeasurementVariableByTermId(TermId.ENTRY_TYPE.getId(), workbook.getFactors()).isPresent();
+
+		// If ENTRY_TYPE variable is not existing in both Condition and Factors Section of workbook
+		// Automatically add ENTRY_TYPE variable as it is required in creating a new Study.
+		if (!entryTypeIdExists) {
+			final MeasurementVariable entryType = this.createMeasurementVariable(TermId.ENTRY_TYPE.getId(), null, Operation.ADD, PhenotypicType.GERMPLASM, programUUID);
+			workbook.getFactors().add(entryType);
+			// Add ENTRY_TYPE variable in Measurement Row with default value Test Entry
+			for (final MeasurementRow row : workbook.getObservations()) {
+				final MeasurementData data = new MeasurementData(entryType.getLabel(), String.valueOf(SystemDefinedEntryType.TEST_ENTRY.getEntryTypeCategoricalId()), false, entryType.getDataType(), entryType);
```

Comment:
Is it possible to create MeasurementData outside of loop? If so we can use forEach clause to simplify the code:
`final MeasurementData data = new MeasurementData(entryType.getLabel(), String.valueOf(SystemDefinedEntryType.TEST_ENTRY.getEntryTypeCategoricalId()), false, entryType.getDataType(), entryType);
			workbook.getObservations().forEach(row -> row.getDataList().add(data));`